src/pass/infer_shape_type.cc

```diff
@@ -74,29 +75,44 @@ Graph InferAttr(Graph &&ret,
           CHECK(is >> rshape[out_ent_id]) << "Invalid attribute";
         }
       }
-    } else if (backward_map.count(inode.source->op())) {
-      // Backward operator inference.
+    } else if (is_backward.get(inode.source->op(), false)) {
       CHECK_GE(inode.control_deps.size(), 1)
         << "BackwardOp need to have control_deps to its forward op";
       const IndexedGraph::Node& fnode = idx[inode.control_deps[0]];
-      // Inference the outputs of backward operator (equal to the inputs
-      // of its corresponding forward operator).
-      std::vector<uint32_t> out_map =
-          backward_map[inode.source->op()](inode.source->attrs);
-      for (size_t i = 0; i < out_map.size(); ++i) {
-        uint32_t in_id = out_map[i];
-        CHECK_LT(in_id, fnode.inputs.size());
-        rshape[idx.entry_id(nid, i)] =
-            rshape[idx.entry_id(fnode.inputs[in_id])];
+      NodePtr fwd_ptr = inode.source->control_deps[0];
+      // use gradient function to find out the correspondence.
+      std::vector<NodeEntry> ograd(fwd_ptr->num_outputs());
+      for (size_t i = 0; i < ograd.size(); ++i) {
+        ograd[i].index = static_cast<uint32_t>(i);
       }
-      if (backward_in_grad.count(inode.source->op())) {
-        std::vector<uint32_t> in_grad =
-          backward_in_grad[inode.source->op()](inode.source->attrs);
-        CHECK_LE(in_grad.size(), fnode.source->num_outputs());
-        for (size_t i = 0; i < in_grad.size(); ++i) {
-          uint32_t eid = idx.entry_id(inode.inputs[in_grad[i]]);
+      // input gradient list
+      auto igrad = fgrad[fwd_ptr->op()](fwd_ptr, ograd);
+      const Op* backward_op = inode.source->op();
+      const Node* igrad_node = nullptr;
+      // Input gradient assignement
+      for (size_t i = 0; i < igrad.size(); ++i) {
+        if (igrad[i].node->op() == backward_op) {
```

Comment:
Does the error point to FGradient?
Is it possible to add a hint after gradianpass calls FGradient